tensorflow/compiler/xla/service/algebraic_simplifier.cc

```diff
@@ -1665,6 +1668,228 @@ StatusOr<HloInstruction*> AlgebraicSimplifierVisitor::OptimizeDotOfGather(
+    lhs_contracting_dims.Add(lhs->dimensions(dim));
+  }
+  lhs = lhs->mutable_operand(0);
+
+  // All checks are passed at this point.
+  //
+  // Transform lhs. Remove the transpose and reshape by sorting the lhs
+  // contracting dims and squish them into a single one. We don't actually
+  // squish the lhs_contracting_dims here because we still need the unsquished
+  // contracting dims to invert reshape and transpose.
+  absl::c_sort(lhs_contracting_dims);
+  lhs = computation_->AddInstruction(
+      HloInstruction::CreateReshape(reshape->shape(), lhs));
+
+  // Transform rhs.
+  // Let's walk through an example to see how rhs transformation is done and the
+  // meaning of some variables used in the implementation. Say we have the
+  // following original Hlo IR:
+  //
+  //   t0 = F32[2, 2, 3] parameter(0)
+  //   t1 = F32[2, 3, 2] transpose(t0) dimensions={0, 2, 1}
+  //   t2 = F32[2, 6] reshape(t1)
+  //   t3 = F32[6, 2] constant(...)
+  //   dot = F32[2, 2] dot(t2, t3) lhs_contracting_dims={1},
+  //                               rhs_contracting_dims={0}
+  //
+  // 1) The sorted lhs_contracting_dims is {1, 2}. We compute t1's subshape
+  //    within the lhs_contracting_dims: reshape_subshape = [3, 2].
+  // 2) rhs_contracting_dims is {0}, t3's shape is [6, 2]. We replace dim 0 of
+  //    t3's shape by reshape_subshape. Now we get inverted reshape's shape:
+  //    inverted_reshape_shape = [3, 2, 2]. We create inverted reshape op:
+  //      inverted_reshape = F32[3, 2, 2] reshape (F32[6, 2] t3)
+  //    and update rhs_contracting_dims to {0, 1}.
+  // 3) We compute the inverse lhs transpose dims
+  //    inverse_transpose_dims = {0, 2, 1}. The lhs contracting dims part is
+  //    inverse_transpose_sub_dims = {2, 1}.
+  // 4) For each element in inverse_transpose_sub_dims, we subtract the start
+  //    position of lhs_contracting_dims, 1 in this example, then add the start
+  //    position of rhs_contracting_dims, 0 in this example. Now we get the
+  //    actual transpose dims within rhs_contracting_dims:
+  //    rhs_transpose_sub_dims = {1, 0}.
+  // 5) We compute the transpose dims of the inverted transpose:
+  //    inverted_transpose_dims = {1, 0, 2}, and then create inverted transpose:
+  //      inverted_transpose =
+  //         F32[2, 3, 2] transpose(inverted_reshape) dimensions = {1, 0, 2}
+
+  // Invert reshape.
+  std::vector<int64> reshape_subshape;
+  reshape_subshape.reserve(lhs_contracting_dims.size());
+  for (auto dim : lhs_contracting_dims) {
+    reshape_subshape.push_back(reshape->operand(0)->shape().dimensions(dim));
+  }
+  CHECK(rhs_contracting_dims.size() == 1);
+  auto inverted_reshape_shape = rhs->shape().dimensions();
+  inverted_reshape_shape.erase(inverted_reshape_shape.begin() +
+                               rhs_contracting_dims[0]);
+  inverted_reshape_shape.insert(
+      inverted_reshape_shape.begin() + rhs_contracting_dims[0],
+      reshape_subshape.begin(), reshape_subshape.end());
```

Comment:
Suppose the original LHS shape was `f32[a,b,c]`, and suppose the squished and transposed LHS shape is `f32[a,cb]`, where `cb` is the contracting dimension.  We want to reorder this to `f32[a,bc]`.

Suppose the original RHS shape is `f32[x,zy]`.  (We never see it in unsquished form.)  We want to unsquish this to `f32[x,z,y]` and then transpose it to `f32[x,y,z]` and finally squish it to `f32[x,yz]`, right?

Moreover, `a == x`, `b == y`, and `c == z`.

Assuming this is right, then can't we compute the unsquished RHS shape as

  - Consider original RHS dimensions
  - Replace RHS contracting dimension with `[c,b]`

?  Like, something as simple as:

```
std::vector<int64> rhs_unsquished_shape_dims = rhs->shape().dimensions();
auto it = rhs_unsquished_shape_dims.erase(rhs_unsquished_shape_dims.begin() + rhs_contracting_dims[0]);
for (int64 dim : lhs_contracting_dims) {
  it = rhs_unsquished_shape_dims.insert(it, lhs->dimensions(dim));
}
```

You'd have to move the `absl::c_sort(lhs_contracting_dims);` bit below here or make a copy or something.

Reply:
A nit, a == x is not necessary, but it does not affect the correctness of this comments. I think the current implementation is more like this:
keep the c_sort as is (so that it is sorted), and
```
std::vector<int64> rhs_unsquished_shape_dims = rhs->shape().dimensions();
auto it = rhs_unsquished_shape_dims.erase(rhs_unsquished_shape_dims.begin() + rhs_contracting_dims[0]);
for (int64 dim : lhs_contracting_dims) {
  it = rhs_unsquished_shape_dims.insert(it, transpose->shape().dimensions(dim));
}
```
The main difference from your comment is that the current impl is getting "b" and "c" from the output of the transpose (i.e. input of the reshape) whereas your comment gets them from input of the transpose. My thought was that since we are inverting reshape op, it would be more natural to look at the shape from original reshape op. But I can go either way.

Reply:
If we are able to reduce the lines of code to 1/3 of what it was and also significantly reduce the amount of comment necessary, that seems like a win to me?

Reply:
I agree. The current code and comments definitely can use some simplification. Let me do an update and we can look at it again. Same for the the inverse transpose case.